File: System/Workspace/Allors.Workspace/ISession.cs
// <copyright file="IDatabase.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System.Threading.Tasks;
    using Data;
    using Meta;
    using System.Collections.Generic;

    public interface ISession
    {
        IWorkspace Workspace { get; }

        ISessionLifecycle SessionLifecycle { get; }

        T Create<T>() where T : class, IObject;

        T Create<T>(IClass @class) where T : IObject;

        T Get<T>(IObject @object) where T : IObject;

        T Get<T>(T @object) where T : IObject;

        T Get<T>(long identity) where T : IObject;

        IEnumerable<T> Get<T>(IEnumerable<IObject> objects) where T : IObject;

        IEnumerable<T> Get<T>(IEnumerable<T> objects) where T : IObject;

        IEnumerable<T> Get<T>(IEnumerable<long> identities) where T : IObject;

        IEnumerable<T> GetAll<T>() where T : class, IObject;

        IEnumerable<T> GetAll<T>(IComposite objectType) where T : IObject;

        void Reset();

        void Merge();

        Task<ICallResult> Call(Method method, CallOptions options = null);

        Task<ICallResult> Call(Method[] methods, CallOptions options = null);

        Task<ILoadResult> Load(params Pull[] pulls);

        Task<ILoadResult> Load(string service, object args);

        Task<ISaveResult> Save();

        IChangeSet Checkpoint();
    }
}

File: System/Workspace/Typescript/libs/workspace/domain/system/src/lib/ISession.ts
import { Composite,  Class} from "@allors/workspace/meta/system";
import { IObject } from './IObject';
import { IWorkspace } from './IWorkspace';
import { Method } from './operands/Method';
import { InvokeOptions } from './database/InvokeOptions';
import { IInvokeResult } from './database/IInvokeResult';
import { IPullResult } from './database/IPullResult';
import { IPushResult } from './database/IPushResult';
import { ISessionLifecycle } from './state/ISessionLifecycle';
import { Pull } from './data/Pull';
import { Procedure } from './data/Procedure';
import { IChangeSet } from './IChangeSet';

export interface ISession {
  workspace: IWorkspace;

  lifecycle: ISessionLifecycle;

  create<T>(cls: Class): T;

  getOne<T>(obj: IObject | T | number | string): T;

  getMany<T>(objects: (IObject | T | number | string)[]): T[];

  getAll<T>(objectType?: Composite): T[];

  invoke(method: Method | Method[], options?: InvokeOptions): Promise<IInvokeResult>;

  pull(pullsOrProceudre: Pull[] | Procedure, pulls?: Pull[]): Promise<IPullResult>;

  push(): Promise<IPushResult>;

  checkpoint(): IChangeSet;

  reset(): void;
}
